src/traffic_cop/traffic_cop.cpp

```diff
@@ -39,9 +38,30 @@
 
 namespace noisepage::trafficcop {
 
+/** The commit callback argument. */
+struct CommitCallbackArg {
+  std::atomic<uint8_t> persist_countdown_;  ///< A countdown latch for what else needs to persist.
+  std::promise<bool> ready_to_commit_;      ///< Set this promise to true to wake up the thread for commit.
+
+  explicit CommitCallbackArg(const transaction::TransactionPolicy &policy) {
+    // The value for the persist_countdown_ field.
+    // Note that the field will be decremented exactly once every time a commit callback is invoked.
+    persist_countdown_ = 0;
+    if (policy.durability_ != transaction::DurabilityPolicy::DISABLE) {
+      persist_countdown_ += 1;
+    }
+    if (policy.replication_ != transaction::ReplicationPolicy::DISABLE) {
+      persist_countdown_ += 1;
+    }
+  }
+};
+
 static void CommitCallback(void *const callback_arg) {
-  auto *const promise = reinterpret_cast<std::promise<bool> *const>(callback_arg);
-  promise->set_value(true);
+  auto *const cb_arg = reinterpret_cast<CommitCallbackArg *const>(callback_arg);
+  --cb_arg->persist_countdown_;
+  if (cb_arg->persist_countdown_ == 0) {
```

Comment:
Is there a race here? I think the comparison should be against a value returned by a post-decrement operator. Otherwise, consider the scenario:

T1: Decrements `persist_countdown_` from 2 to 1
T2: Decrements `persist_countdown_` from 1 to 0. Reads `persist_countdown_` as 0, calls `set_value()`.
T1: Reads `persist_countdown_` as 0, calls `set_value()`.

It's possible that `std::promise::set_value` is idempotent, but I didn't spot that property in cppreference so the behavior of calling it multiple times is unclear to me. Maybe it generates an exception?

Reply:
yea, you're right! good catch